### DISCUSSION TITLE
Contributing guidelines (`CONTRIBUTING.{rst,md}`)

### DISCUSSION THREAD

**Discussion by aucampia** [RDFLib/rdflib#1912] [Development]
Currently we have nothing in place that GitHub recognizes as [contributing guidelines](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors). In addition to this we don't exactly have any guidelines describing how someone can contribute, specifically these guidelines should include things like:

- Review pending pull requests marked with review wanted
- Add additional tests for untested code
- Modernize test files
- Fixed expected test failures (i.e. known issues)
- Add/fix docstrings
- Add more expected failure tests for issues that are not yet fixed

These two problems are not entirely the same, but they are closely related.

Some options to consider:

1. Refashion our developer guide into contributing guidelines and add the additional points on how to contribute, and also do what is needed to make GitHub recognize it.
2. Write a new contribution guidelines document and link from that to our developers guide.

I'm very split on which to do, I am slightly leaning towards 1. but I would like to hear more feedback before moving forward with it.

CC: @RDFLib/core

**Comment by aucampia** [comment]
Some more things to add:
- Update wikipedia page https://en.wikipedia.org/wiki/RDFLib
- Update wikidata entries https://www.wikidata.org/wiki/Q7276224
- Help out on stackoverflow https://stackoverflow.com/questions/tagged/rdflib
- Help out on gitter https://gitter.im/RDFLib/rdflib
- Help out in discussions https://github.com/RDFLib/rdflib/discussions

**Comment by aucampia** [comment]
I have thought about this some more, I tihnk I'm going to go with the second option as a start, I will try make a PR for this before the start of 2022W21.

**Reply by aucampia** [reply]
I made a PR for this:
- https://github.com/RDFLib/rdflib/pull/2082

**Comment by ghost** [comment]
I need a bit more time before I can substantively comment on/contribute to this. I believe there's quite a lot to be addressed here and I want to take a pass through the closed PRs to get a sense of to what degree they relate to specific items in the developer guide.

I'm currently absorbing the findings reported in:

1. [“Report on the 2020 FOSS Contributor Survey”](https://www.linuxfoundation.org/wp-content/uploads/2020FOSSContributorSurveyReport_121020.pdf) - informative overview
2. [“Understanding Contributor Behavior within Large Free/Open Source Software Projects: A Socialization Perspective”](https://www.researchgate.net/profile/Kevin-Carillo-2/publication/271522003_Understanding_Contributor_Behavior_within_Large_FreeOpen_Source_Software_Projects_A_Socialization_Perspective/links/54cb50070cf22f98631e7244/Understanding-Contributor-Behavior-within-Large-Free-Open-Source-Software-Projects-A-Socialization-Perspective.pdf) - useful characterisation
3. [“The Signals that Potential Contributors Look for When Choosing Open-source Projects”](https://dl.acm.org/doi/pdf/10.1145/3359224)  - rather relevant

**Comment by ghost** [comment]
And I thought this issue template might be a reasonable starting point, although IMO, it needs “informalizing”:
```
<!-- This issue tracker is only for technical issues related to RDFLib.

General RDFLib questions and/or support requests are best directed to Stackoverflow with the [rdflib] tag.

-->

<!-- Describe the issue -->
<!--- What behaviour did you expect? -->

<!--- What was the actual behaviour? -->

<!--- How reliably can you reproduce the issue and what are the steps to do so? -->

<!-- What version of RDFLib are you using? -->

<!-- What version of Python are you using? -->

<!-- What type of machine are you observing the error on (OS/CPU and disk type)? -->

<!-- Any extra information that might be useful in the debugging process. -->
<!--- Raw text or a link to a pastebin type site are preferred. -->
```
